check_well_founded: give expansion traces in error messages

gasche opened the pull request:
I just restarted reviewing #11207 (a PR which fixes a compiler-performance bug). I'm convinced that the new implementation is correct, but one aspect of the change that is fairly tricky to reason about is its effect on error messages. It seems to make some error messages worse, but it's not clear why.

I propose to make the difference (in quality of the error messages between the two algorithms to check well-foundedness of type definitions) obvious by making the error messages informative in the first place. The present PR therefore gives a full trace of expansions that lead to a cycle, instead of just giving (sometimes) the final type we reach during cycle detection. I believe that this will be useful to review this aspect of #11207, but note that this is already a usability improvement with today's algorithm, independently of the other PR.

Note: with the current way that well-foundedness is checked, there is in fact little difference between the two different exceptions, Recursive_abbrev and Cycle_in_def. (Recursive_abbrev sounds a bit more confident in blaming the cycle precisely on a given definition, but otherwise they are explained in the same way.) For now I would suggest keeping the two distinct errors instead of merging them, to see whether they remain similar or grow more different with the different algorithm in #11207.

(cc @garrigue @Octachron)

gasche committed:
check_well_founded: give expansions traces in error messages

Note: with the current way that well-foundedness is checked, there is
in fact little difference between the two different exceptions,
Recursive_abbrev and Cycle_in_def. (Recursive_abbrev sounds a bit more
confident in blaming the cycle precisely on a given definition, but
otherwise they are explained in the same way.) For now we suggest
keeping the two distinct errors instead of merging them, to see
whether they remain similar or grow more different with the different
algorithm in #11207.

gasche commented on testsuite/tests/typing-poly/poly.ml:
```diff
@@ -1025,8 +1026,7 @@ and  ('a2, 'b2) ty2 = 'b2 -> unit constraint 'b2 = [> `V2 of ('a2, 'b2) ty1 as '
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Error: The definition of ty1 contains a cycle:
-       [> `V1 of ('a, 'b) ty2 as 'b ] as 'a
+Error: The definition of ty1 contains a cycle
```
This example gives an arguably slightly worse message than before. (Other cases are clear improvements, in my opinion.) We could get the best of both worlds by keeping the "last type informed in the cycle" parameter and printing it only in case the expansion sequence is omitted (as in this case) because it is too short. But we would also get the code complexity of both worlds, and I think that it is not worth it just to get a slight improvement in those pathological examples involving constraints.

garrigue replied:
In the standard cases, where a cycle can be easily exhibited, this PR is a clear improvement.
On the other hand, in some harder cases where the cause is more subtle, there is no output at all.
To me this looks like it may make debugging of the compiler harder, since we cannot even point where the problem happens.
I didn't look much at the code, but would it possible to print a shorter trace, if length is the problem?

gasche replied:
Thanks for your feedback. I will try to tweak the PR to always show at least this "final" type.

gasche replied:
The new output on this example, after I rebased the PR to also trace subterm relations:

```
Error: The definition of ty1 contains a cycle:
         [> `V1 of ('a, 'b) ty2 as 'b ] as 'a contains 'b
```

gasche commented on testsuite/tests/typing-objects/Tests.ml:
```diff
@@ -222,16 +222,17 @@ and 'a t = 'a t u;;
-Error: The type abbreviation t is cyclic
+Error: The type abbreviation t is cyclic:
+         'a t = 'a t u,
+         'a t = 'a t u
```
This example and the one below look a bit strange. This comes from the fact that the cycle-detection logic compares the identity of the node of the type-graph, not their structure. The two occurrences of `'a t` here look similar but they are distinct type nodes, so `'a t = 'a t u` is not immediately found to be cyclic. I think that we need to expand `'a t` again to eventually encounter a type whose identity you have already encountered during the traversal.

This is slightly weird for users, but also fairly useful for the people trying to understand the cycle-detection algorithm, and I think not harmful for users either.

gasche replied:
The new output on this example:

```
Error: The type abbreviation t is cyclic:
         'a t = 'a t u,
         'a t u contains 'a t,
         'a t = 'a t u,
         'a t u = 'a t
```

The same reasoning applies, but it is now easier to follow what is going on.

gasche commented:
@garrigue I updated this PR to address your concern

> In the standard cases, where a cycle can be easily exhibited, this PR is a clear improvement.
On the other hand, in some harder cases where the cause is more subtle, there is no output at all.
To me this looks like it may make debugging of the compiler harder, since we cannot even point where the problem happens.

The cause for having less output in some cases was that I would only trace expansions, and not the subterm relation, while both kind of steps are taken during our cycle-detection (or regularity-checking algorithms).

The new version of the PR has traces that contain both kind of steps. This is sometimes a bit redundant, but the output is now always more information than what we had before.

```
type 'a t = 'a * 'b constraint 'a = 'b t;;
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: The type abbreviation t is cyclic:
         'a t t = 'a t * 'a,
         'a t * 'a contains 'a t,
         'a t = 'a t * 'a
```

I also changed the format of the "regularity" traces that pre-existed; this adds more details to some error messages.

```
type 'a c = <f : 'a c; g : 'a d>
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
and 'a d = <f : int c>;;
Error: This recursive type is not regular.
       The type constructor c is defined as
         type 'a c
       but it is used as
         int c
       after the following expansion(s):
         < f : 'a c; g : 'a d > contains 'a d,
         'a d = < f : int c >,
         < f : int c > contains int c
       All uses need to match the definition for the recursive type to be regular.
```

We can decide later to elide some of this information. But a good part of the redundancy we see in the testsuite output in this PR comes from suboptimal aspects of the current cycle-detection algorithm (so: it reveals a deeper issue), and merging the simpler algorithm from #11207 would improve things. To get a sense of the improved, I rebased #11207 on top of this branch, to see the changes:

The combination of this PR with #11207: https://github.com/ocaml/ocaml/compare/trunk...gasche:ocaml:fix11150-error-message
The changes of #11207 alone, but on top of this PR: https://github.com/ocaml/ocaml/commit/0a6809a3e83aafe96f953b450048611d65c80bdc

In particular, looking at testsuite changes in https://github.com/ocaml/ocaml/commit/0a6809a3e83aafe96f953b450048611d65c80bdc we can see that #11207 often simplifies/improves the expansion sequence, which is good information. If we want to rediscuss aspects of #11207 or further experiment, this expansion output could be very useful.

My proposal remains to merge the current PR (#11722) if it is acceptable, and then rebase the more complex #11207 on top of it. (The changes to implement trace collection are simple and systematic, so rebasing is easy.)

garrigue reviewed (approved):
This looks much better indeed.
The stuttering is a bit confusing to users, but it is hard to avoid as types that are semantically equal do not necessarily share the same type node: inside cycles, they will eventually do, but this may take a while.

gasche commented:
I had an idea for the repetition (I would call these "repetitions" rather than "stuttering" because the repeated expansions or types are typically not consecutive): after we detect the cycle and collect the trace, we could post-process the trace to check a richer notion of structural equality between nodes (not just physical equality) and cut the trace earlier when possible.
This would be a nice-to-have. I propose to merge the PR right away, but I may send a follow-up PR if the idea works.

gasche commented:
Note: I tried the obvious thing using `Ctype.is_equal`, and the implementation seems to loop forever on cyclic types.

Edit: reading the code, it looks like the function is doing the right thing on equi-recursive types (it carries a set of pairs of types already visited), but it assumes (reasonably!) that type abbreviations are well-founded and expands them eagerly, which loops on our definitions that are precisely not well-founded.

gasche commented:
@garrigue my current best attempt to implement this idea is at

https://github.com/gasche/ocaml/tree/typedecl-shorter-expansion-traces

this seems to work very nicely in some cases and fail horribly in some others. I have no idea why it fails horribly, and (this is not the priority but) I would be curious to understand why. Quoting the commit message:

> This version appears to:
> 
> 1. produce shorter error traces most of the time, as expected
> 
> 2. sometimes *modify* the types printed, which is unexpected
> 
> 3. loop infinitely on
>     typing-objects/Tests.ml
>     typing-poly/poly.ml
>    which is also unexpected (probably a bug?)
> 
> 4. sometimes return *much longer* expansions for certain types,
>    which is *very* unexpected. I don't know how this is possible,
>    except possibly by some in-place modification (2) caused by
>    a previous call in the same toplevel session?
